build_kernel.sh

```diff
@@ -10,62 +11,102 @@ OUTPUT_DIR=$DIR/output
+  echo "Cloning agnos-kernel-sdm845"
   git submodule update --init agnos-kernel-sdm845
 fi
-cd agnos-kernel-sdm845
 
-# Build parameters
-ARCH=$(uname -m)
-if [ "$ARCH" != "arm64" ] && [ "$ARCH" != "aarch64" ]; then
-  export CROSS_COMPILE=$TOOLS/aarch64-linux-gnu-gcc/bin/aarch64-linux-gnu-
-  export CC=$TOOLS/aarch64-linux-gnu-gcc/bin/aarch64-linux-gnu-gcc
-  export LD=$TOOLS/aarch64-linux-gnu-gcc/bin/aarch64-linux-gnu-ld.bfd
+$DIR/tools/extract_tools.sh
 
-  $DIR/tools/extract_tools.sh
-fi
-export ARCH=arm64
-
-# these do anything?
-export KCFLAGS="-w"
-
-# Load defconfig and build kernel
-echo "-- First make --"
-make $DEFCONFIG O=out
-echo "-- Second make: $(nproc --all) cores --"
-make -j$(nproc --all) O=out  # Image.gz-dtb
-
-# Turn on if you want perf
-# LDFLAGS=-static make -j$(nproc --all) -C tools/perf
-
-# Copy over Image.gz-dtb
-mkdir -p $TMP_DIR
-cd $TMP_DIR
-cp $DIR/agnos-kernel-sdm845/out/arch/arm64/boot/Image.gz-dtb .
-
-# Make boot image
-$TOOLS/mkbootimg \
-  --kernel Image.gz-dtb \
-  --ramdisk /dev/null \
-  --cmdline "console=ttyMSM0,115200n8 quiet loglevel=3 earlycon=msm_geni_serial,0xA84000 androidboot.hardware=qcom androidboot.console=ttyMSM0 ehci-hcd.park=3 lpm_levels.sleep_disabled=1 service_locator.enable=1 androidboot.selinux=permissive firmware_class.path=/lib/firmware/updates net.ifnames=0 dyndbg=\"\"" \
-  --pagesize 4096 \
-  --base 0x80000000 \
-  --kernel_offset 0x8000 \
-  --ramdisk_offset 0x8000 \
-  --tags_offset 0x100 \
-  --output $BOOT_IMG.nonsecure
-
-# le signing
-openssl dgst -sha256 -binary $BOOT_IMG.nonsecure > $BOOT_IMG.sha256
-openssl pkeyutl -sign -in $BOOT_IMG.sha256 -inkey $DIR/vble-qti.key -out $BOOT_IMG.sig -pkeyopt digest:sha256 -pkeyopt rsa_padding_mode:pkcs1
-dd if=/dev/zero of=$BOOT_IMG.sig.padded bs=2048 count=1
-dd if=$BOOT_IMG.sig of=$BOOT_IMG.sig.padded conv=notrunc
-cat $BOOT_IMG.nonsecure $BOOT_IMG.sig.padded > $BOOT_IMG
-
-# Copy to output dir
-mkdir -p $OUTPUT_DIR
-mv $BOOT_IMG $OUTPUT_DIR/
-cp $DIR/agnos-kernel-sdm845/out/techpack/audio/asoc/snd-soc-sdm845.ko $OUTPUT_DIR/
-cp $DIR/agnos-kernel-sdm845/out/techpack/audio/asoc/codecs/snd-soc-wcd9xxx.ko $OUTPUT_DIR/
-cp $DIR/agnos-kernel-sdm845/out/drivers/staging/qcacld-3.0/wlan.ko $OUTPUT_DIR/
+build_kernel() {
+  cd agnos-kernel-sdm845
+
+  # Build parameters
+  ARCH=$(uname -m)
+  if [ "$ARCH" != "arm64" ] && [ "$ARCH" != "aarch64" ]; then
+    export CROSS_COMPILE=$TOOLS/aarch64-linux-gnu-gcc/bin/aarch64-linux-gnu-
+    export CC=$TOOLS/aarch64-linux-gnu-gcc/bin/aarch64-linux-gnu-gcc
+    export LD=$TOOLS/aarch64-linux-gnu-gcc/bin/aarch64-linux-gnu-ld.bfd
+  fi
+
+  # Build arm64 arch
+  export ARCH=arm64
+
+  # Set ccache dir
+  export CCACHE_DIR=$DIR/.ccache
+
+  # Avoid LINUX_COMPILE_HOST to change on every run thus invalidating cache
+  # https://patchwork.kernel.org/project/linux-kbuild/patch/1302015561-21047-8-git-send-email-mmarek@suse.cz/
+  export KBUILD_BUILD_HOST="docker"
```

Comment:
@adeebshihadeh the issue with increasing cached build time in docker was because LINUX_COMPILE_HOST had the container id (which is different everytime). setting KBUILD_BUILD_HOST="docker" makes LINUX_COMPILE_HOST="docker" (instead of the container id).